File: program-native/src/program_command.rs
use simple_serde::SimpleSerde;

#[repr(C)]
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum Command {
    InitDashboard, // Initialize a dashboard account
    InitPlayer,    // Initialize a player account
    InitGame,      // Initialize a game account

    Advertise,    // Used by Player X to advertise their game
    Join,         // Player O wants to join
    KeepAlive,    // Player X/O keep alive
    Move(u8, u8), // Player X/O mark board position (x, y)
}
impl SimpleSerde for Command {}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    pub fn serialize() {
        let cmd = Command::InitDashboard;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::InitPlayer;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::InitGame;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::Advertise;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::Join;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::KeepAlive;
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        let cmd = Command::Move(1, 2);
        let mut b = vec![0; 16];
        cmd.serialize(&mut b).unwrap();
        assert_eq!(b, [6, 0, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
    }
}

File: src/program/program-command.js
/**
 * The commands (encoded as Transaction Instructions) that are accepted by the
 * TicTacToe Game and Dashboard program
 *
 * @flow
 */

import * as BufferLayout from 'buffer-layout';

const COMMAND_LENGTH = 6;

const Command = {
  InitDashboard: 0, // Initialize a dashboard account
  InitPlayer: 1, // Initialize a player account
  InitGame: 2, // Initialize a game account
  Advertise: 3, // Used by Player X to advertise their game
  Join: 4, // Player O wants to join
  KeepAlive: 5, // Player X/O keep alive
  Move: 6, // Player X/O mark board position (x, y)
};

function zeroPad(command: Buffer): Buffer {
  if (command.length > COMMAND_LENGTH) {
    throw new Error(
      `command buffer too large: ${command.length} > ${COMMAND_LENGTH}`,
    );
  }
  const buffer = Buffer.alloc(COMMAND_LENGTH);
  command.copy(buffer);
  return buffer;
}

function commandWithNoArgs(command: number): Buffer {
  const layout = BufferLayout.struct([BufferLayout.u32('command')]);
  const buffer = Buffer.alloc(layout.span);
  layout.encode({command}, buffer);
  return zeroPad(buffer);
}

export function initDashboard(): Buffer {
  return commandWithNoArgs(Command.InitDashboard);
}

export function initPlayer(): Buffer {
  return commandWithNoArgs(Command.InitPlayer);
}

export function initGame(): Buffer {
  return commandWithNoArgs(Command.InitGame);
}

export function advertiseGame(): Buffer {
  return commandWithNoArgs(Command.Advertise);
}

export function joinGame(): Buffer {
  return commandWithNoArgs(Command.Join);
}

export function keepAlive(): Buffer {
  return commandWithNoArgs(Command.KeepAlive);
}

export function move(x: number, y: number): Buffer {
  const layout = BufferLayout.struct([
    BufferLayout.u32('command'),
    BufferLayout.u8('x'),
    BufferLayout.u8('y'),
  ]);

  const buffer = Buffer.alloc(layout.span);
  layout.encode({command: Command.Move, x, y}, buffer);
  return zeroPad(buffer);
}
